### ISSUE TITLE
TMB error when predicting with se.fit=TRUE

### ISSUE THREAD

**Issue by marek-tph** [glmmTMB/glmmTMB#545]
When trying to predict with standard errors from a fitted model (mixed effects beta regression), like so:
```
predict(m, newdata=df, se.fit=TRUE)
```
I often get the following error message:
```
Error in if (all(Dphi.random == 0)) { : 
  missing value where TRUE/FALSE needed
```
I am unable to provide a reproducible example short of uploading my whole data, because I can't isolate the error in any way. It appears for some models but not others, and for some datasets and not others (notably, it does not appear when predicting on the training data). I tracked down the relevant piece of TMB code, I believe:  https://github.com/cran/TMB/blob/master/R/sdreport.R#L205, but I simply don't know what it does. All I can deduce is that somewhere upstream from this point NAs were generated but were not handled.

I don't expect a complete resolution of the problem given the vagueness of the report, but an educated guess at a root cause would be very welcome, so that I can figure out a workaround. 

Thank you in advance, and also for fixing #533.

**Comment by bbolker** [comment]
At a glance I'm not super-familiar with this bit of code either.  It would be easy enough to put in a test here so that the code didn't stop (e.g. just propagated `NA` values for the standard errors), but I'm not sure that would constitute a "workaround".  **Any** reproducible example (data shared privately if necessary) would make this a lot easier to debug ...

**Comment by marek-tph** [comment]
I had another look with fresh eyes, and the problem appears when there are `NA`s in fixed covariates in the prediction data. Consider

```
library(glmmTMB)

groups <- c("a", "b", "c", "d", "e", "f", "g", "h", "i", "j")
effs <- rnorm(10, sd=0.1)
x <- rnorm(100) + rep(effs, 10)
y <- exp(x)/(1 + exp(x)) + rnorm(100, sd=0.01)
g <- rep(groups, 10)

m1 <- glmmTMB(y ~ x, family=beta_family())
m2 <- glmmTMB(y ~ x + (1|g), family=beta_family())
```
This sets up 100 observations across 10 groups and fits two models, with and without a group (random) effect. Now let's try predicting with an extra observation where the fixed predictor is `NA`:

```
> df <- data.frame(x=c(NA, x), g=c("a", g), stringsAsFactors=FALSE)
> 
> head(as.data.frame(predict(m1, newdata=df, se.fit=TRUE)))
         fit se.fit
1         NA     NA
2  1.2273146     NA
3  0.5216004     NA
4 -1.6377053     NA
5  0.6593653     NA
6  0.6297232     NA
```
This is arguably already iffy, since one would expect SEs to be computed for the unproblematic observations. When the model has a random effect, the original error appears:

```
> head(as.data.frame(predict(m2, newdata=df, se.fit=TRUE)))
Error in if (all(Dphi.random == 0)) { : 
  missing value where TRUE/FALSE needed
```

Predicting over `NA`s is dumb and I should have realised what was happening before opening a vague issue, but I think that the user should get a specific error message in such cases.

**Comment by mebrooks** [comment]
I think I added a more informative error message about this in the development version. Sorry I can’t look into it at this moment, but I will later.

**Comment by bbolker** [comment]
It should be certainly be possible in principle to add an `na.action` argument to `predict.glmmTMB` and use `napredict()` within the function to restore `NA` values for predictions where the original input has `NA` values in the fixed effects.  `lme4` has some very ugly code for handling all of these cases; it works, but I think a *careful* enumeration of all the possibilities (`NA` in fixed effects, `NA` in random effects, etc.) would be worthwhile before jumping in.

**Comment by bbolker** [comment]
PS I can confirm that in the development version, either case gives:

> Error in eval(substitute(expr), data, enclos = parent.frame()) : 
  Some variables in newdata needed for predictions contain NAs or NaNs.
           This is currently incompatible with se.fit=TRUE.

**Comment by dylangomes** [comment]
I am running into the same error with no NAs in the dataset.

**Comment by bbolker** [comment]
Can we have a reproducible example please ...?

**Comment by dylangomes** [comment]
Yes, I have just created a simulated dataset, as a reproducible example, but the error does not occur when I do that. For some reason it still occurs with my dataset (code and csv here: [https://github.com/dylangomes/predict_glmmTMB]). The only difference I can think of is that all of my variables, except one, are set to not vary - set to the mean. Is this an issue?

**Comment by bbolker** [comment]
it could be.  What if you simulate data with that characteristic?

**Comment by dylangomes** [comment]
hmm, no that seems to run just fine. In the real data set, the predict function only seems to work with one variable in the model at a time, yet predict doesn't seem to have a problem with models containing more than 1 variable if the source is simulated. The structure of the data are as they should be, all numeric and one factor, and no NAs. I have a custom scale function within the models, which I thought may be an issue, but predict.glmmTMB has no issue with one predictor variable that is wrapped in this function.
